src/repr/src/row.rs

```diff
@@ -1848,13 +2141,214 @@ mod tests {
+            // Validates Ord implementation; relies on input being hand-sorted.
+            assert!(last_seen < row_value, "{last_seen} < {row_value}");
+            test_ranges_inner(row_value, vec_value, string_rep);
+            last_seen = row_value;
+        }
+
+        // Test null handling
+        for (lower, upper, s) in [
+            // This essentially tests the encoding we'll need for non-partition
+            // reclocking data.
+            (None, Some(Datum::Null), "(,null)"),
+            (Some((Datum::Null, true)), None, "[null,)"),
+            (Some((Datum::Null, false)), None, "(null,)"),
+        ] {
+            let r = arena.make_datum(|packer| {
+                let (lower, lower_inc) = match lower {
+                    None => (None, false),
+                    Some((d, lower_inc)) => (Some(d), lower_inc),
+                };
+                packer
+                    .push_range_with(
+                        RangeBoundDesc {
+                            inf: lower.is_none(),
+                            inc: lower_inc,
+                        },
+                        RangeBoundDesc {
+                            inf: upper.is_none(),
+                            inc: false,
+                        },
+                        |row| {
+                            for bound in [lower, upper] {
+                                if let Some(bound) = bound {
+                                    row.push(bound);
+                                }
+                            }
+                        },
+                    )
+                    .unwrap();
+            });
+            // We don't care about equality between the two, but do want to
+            // test roundtripping.
+            test_ranges_inner(r, &r, &s);
+            datum_vec.push((r, s));
+        }
+    }
+
+    fn test_range_errors_inner<'a>(datums: Vec<Datum<'a>>) -> Result<(), RangeError> {
+        let mut row = Row::default();
+        let row_len = row.byte_len();
+        let mut packer = row.packer();
+        let r = packer.push_range_with(
+            RangeBoundDesc {
+                inf: false,
+                inc: false,
+            },
+            RangeBoundDesc {
+                inf: false,
+                inc: false,
+            },
+            |row| {
+                for v in datums {
+                    row.push(v);
+                }
+            },
+        );
+
+        assert_eq!(row_len, row.byte_len());
+
+        r
+    }
+
+    macro_rules! test_range_errors_panic {
+        ($(($nonce:expr, $val:expr)),+) => {
+            $(
+                paste::item! {
+                    #[test]
+                    #[should_panic]
+                    fn [< test_range_errors_panic_ $nonce>] () {
+                        let _ = test_range_errors_inner($val);
+                    }
+                }
+            )+
+        };
+    }
```

Comment:
This is not worth the macro, IMO! Macros have a fairly high cognitive overhead for the reader so I try to bias for not using them unless they are sparing the writer from serious tedium. In this case with just three examples I think it actually works out to less lines of code to copy/pasta.

If you wanna stick with the table driven style, you might be interested in `mz_ore::panic::catch_unwind`? I usually find that nicer to work with than the `should_panic` macro!

Reply:
I hadn't known about `catch_unwind`; solves all my problems. Ty.